classes-recipe/platsch.bbclass

```diff
@@ -0,0 +1,84 @@
+# Class to simplify the conversion of images to be displayed by "platsch" as splash images
+#
+# Description:
+#
+# In the recipe list the conventionally formatted images that shall be converted
+# and specify the raw format that they shall be converted to:
+#
+#   SPLASH_IMAGES = "splash-1280x800.png splash-800x480.png"
+#   SPLASH_FORMAT = "RGB565"
+#
+# Of course, those source images have to be listed in SRC_URI beforehand.
+#
+# If the raw format that you need for your hardware is not supported by this
+# class up to now, you are invited to create a corresponding pull request.
+#
+# If the resulting images shall not be installed to the default location (which
+# is /usr/share/platsch) choose another directory via
+#
+#   SPLASH_DATADIR = "/some/different/directory"
+#
+# Finally the class automatically appends the converted images to the
+# FILES:${PN} parameter.
+#
+
+SPLASH_IMAGES ??= ""
+SPLASH_IMAGES[doc] = "Space separated list of conventionally formatted images (*.png, *.jpg etc.) to be converted by this class."
+
+SPLASH_FORMAT ??= "undefined"
+SPLASH_FORMAT[doc] = "Target format that the graphics hardware's display controller is able to scan out."
+
+SPLASH_DATADIR ??= "${datadir}/platsch"
+SPLASH_DATADIR[doc] = "Directory where the converted images will be installed."
+
+DEPENDS = "imagemagick-native"
+
+# Remove and re-create ${B} so that it is guaranteed to be empty
+do_compile[cleandirs] = "${B}"
+
+platsch_do_compile() {
+    for IMG in ${SPLASH_IMAGES}; do
+        echo $IMG | sed -E -e "s/(.*)-(.*)x(.*)\.(.*)/\2 \3 \4/" |
+        {
+            read WIDTH HEIGHT EXT
+            BASENAME=`basename --suffix=".${EXT}" ${IMG}`
+
+            case ${SPLASH_FORMAT} in
+            RGB565)
+                # HACK: a quite compact way to create a bitmap in the RGB565 format is to
+                # let ImageMagick convert the source into its known BMP subtype RGB565 and
+                # to strip the BMP header by only using the final width * height * (5+6+5)/8
+                # bytes. Explicitly request an RGB format to avoid color map and alpha
+                # channel (via asking for the TrueColor type, see [1]).
+                #
+                # [1] https://imagemagick.org/script/command-line-options.php#type
+                #
+                RGB565_SIZE=`expr ${WIDTH} "*" ${HEIGHT} "*" \( 5 + 6 + 5 \) / 8`
```

Comment:
The fact that image width and height are expected in the file name is only implicitly documented. I think it should be possible to get this information via `magick identify`, so we don't need to rely on this. I don't have a strong opinion whether that's too implicit or not.

Since the `magick` command doesn't do any resizing, this would also prevent specifying wrong dimensions in the file name.

Reply:
Right, I completely forgot to describe that the class expects images with filenames following a prescribed scheme - fixed that [here](https://github.com/pengutronix/meta-ptx/compare/3171c8f5ec692b745354d004adb95390353533eb..417d3c0c629b4f4de62dd9db27fb62a4d7c200ac#diff-b15ffc8dbe2d6e03cc3d405e3325956193c6010e487b7d4db01d6d6f1fd63f4bR5-R11) and [here](https://github.com/pengutronix/meta-ptx/compare/3171c8f5ec692b745354d004adb95390353533eb..417d3c0c629b4f4de62dd9db27fb62a4d7c200ac#diff-b15ffc8dbe2d6e03cc3d405e3325956193c6010e487b7d4db01d6d6f1fd63f4bR48) in the updated version. Additionally the code now checks for [``WIDTH, HEIGHT & EXT``](https://github.com/pengutronix/meta-ptx/compare/3171c8f5ec692b745354d004adb95390353533eb..417d3c0c629b4f4de62dd9db27fb62a4d7c200ac#diff-b15ffc8dbe2d6e03cc3d405e3325956193c6010e487b7d4db01d6d6f1fd63f4bR52) as well as [``BASENAME``](https://github.com/pengutronix/meta-ptx/compare/3171c8f5ec692b745354d004adb95390353533eb..417d3c0c629b4f4de62dd9db27fb62a4d7c200ac#diff-b15ffc8dbe2d6e03cc3d405e3325956193c6010e487b7d4db01d6d6f1fd63f4bR66) to not be empty and so implicitely that the filename satisfies the given scheme.

To prevent that the geometry encoded in the filename and the geometry found in the image do not match a [check](https://github.com/pengutronix/meta-ptx/compare/3171c8f5ec692b745354d004adb95390353533eb..417d3c0c629b4f4de62dd9db27fb62a4d7c200ac#diff-b15ffc8dbe2d6e03cc3d405e3325956193c6010e487b7d4db01d6d6f1fd63f4bR56) has been added, that utilizes ImageMagick's ``identify`` to extract the needed information.

If one of those checks fails, the class is left via ``bbfatal`` as this is something that the user really wants to fix.

In accordance with ``platsch``'s [README](https://git.pengutronix.de/cgit/platsch/tree/README.rst#n53) the call to ``magick`` to finally do the conversion got an additional command line option [``-strip``](https://imagemagick.org/script/command-line-options.php#strip) to to get rid of any profiles, comments and several PNG chunks that perhaps could interfere.

Additionally I did some minor cleanup:
- replaced the old backquote form of command substitution `` `<...>` `` by the modern ``$(<...>)``
- escaped all mathematical symbols in the call of ``expr`` by double quotes
- delimited the text printed by ``bbfatal`` in double quotes